### PR TITLE
WIP Owls104087 (solution 1) fix an intermitent failure in ItKubernetesDomainEvents#testK8SEventsDelete

### PR DESCRIPTION
**Pull request by doxiao**
The PR fixes an intermittent issue in nightly where the ClusterDeleted event is missing when deleting a domain and cluster resources.

- Generate ClusterCreated/Deleted/Changed events even if the cluster is not referenced by any domain. 
- When a cluster is referenced by a domain, the events are generated in a cluster-event-only MakeRightOperation followed by a regular MakeRightOperation to do the necessary domain processing.

(https://build.weblogick8s.org:8443/job/weblogic-kubernetes-operator-kind-new/14141)
(https://build.weblogick8s.org:8443/job/weblogic-kubernetes-operator-kind-new/14150)
https://build.weblogick8s.org:8443/job/weblogic-kubernetes-operator-kind-new/14224/

### CODE REVIEW

**Review by alai8 — APPROVED** [review]

**Comment by ankedia** [comment]
>MakeRightOperation for deleting a domain waits until the current in-flight MakeRightOperation for the same domain completes.

Does this mean MakeRightOperation for deleting a domain will also wait in other cases such as rolling restart or introspection? I think we won't want to wait for a roll or introspection to complete if the domain is deleted.

**Comment by rjeberhard** [comment]
> > MakeRightOperation for deleting a domain waits until the current in-flight MakeRightOperation for the same domain completes.
> 
> Does this mean MakeRightOperation for deleting a domain will also wait in other cases such as rolling restart or introspection? I think we won't want to wait for a roll or introspection to complete if the domain is deleted.

@doxiao, I'm going to be interested in this answer too. It would be inappropriate, unfortunately, to wait for a make-right to complete (e.g. if that includes starting up servers) if the domain has been deleted.

File: operator/src/main/java/oracle/kubernetes/operator/DomainProcessorImpl.java [hunk 1]
```diff
@@ -697,42 +705,76 @@ public void dispatchClusterWatch(Watch.Response<ClusterResource> item) {
...
-          .interrupt()
-          .withExplicitRecheck()
-          .withEventData(new EventData(EventItem.CLUSTER_CREATED).resourceName(cluster.getClusterName()))
-          .execute();
-    });
+    List<DomainPresenceInfo> hostingDomains =
+        getExistingDomainPresenceInfoForCluster(cluster.getNamespace(), cluster.getMetadata().getName());
+    if (hostingDomains.isEmpty()) {
+      LOGGER.info(MessageKeys.WATCH_CLUSTER_WITHOUT_DOMAIN, cluster.getMetadata().getName());
+      createMakeRightOperationForClusterEvent(EventItem.CLUSTER_CREATED, cluster).execute();
```
Comment (ankedia) [inline operator/src/main/java/oracle/kubernetes/operator/DomainProcessorImpl.java:712]: I wonder if we may trigger a make-right for some stranded cluster to generate the CLUSTER events and make-right might use some stale DPI.

Comment (doxiao) [reply operator/src/main/java/oracle/kubernetes/operator/DomainProcessorImpl.java:712]: Can you elaborate on the benefit of doing that?

**Comment by doxiao** [comment]
> > MakeRightOperation for deleting a domain waits until the current in-flight MakeRightOperation for the same domain completes.
> 
> Does this mean MakeRightOperation for deleting a domain will also wait in other cases such as rolling restart or introspection? I think we won't want to wait for a roll or introspection to complete if the domain is deleted.

yes, the MakeRightOperation waits on any in-flight operation to the same domain right now.

File: operator/src/main/java/oracle/kubernetes/operator/DomainProcessorImpl.java (hunk 1 above)
Comment (ankedia) [reply operator/src/main/java/oracle/kubernetes/operator/DomainProcessorImpl.java:712]: Sorry, I was not suggesting to do that but meant to ask a question. We can discuss this in the offline slack thread.

**Review by lennyphan — APPROVED** [review]

**Review by rjeberhard — APPROVED** [review]

**Pull request closed without merging**